processor/src/host/debug.rs

```diff
@@ -1,52 +1,242 @@
     }
 
-    // print the last item, and in case the stack has more items, print the total number of
-    // un-printed items
-    let i = num_items - 1;
-    if num_items == stack.len() {
-        println!("└── {i:>2}: {}", stack[i]);
+    // print the last memory value
+    if let Some((addr, value)) = mem_interval.last() {
+        print_mem_address(*addr, *value, true, is_local, padding);
+    }
+}
+
+/// Prints single memory value with its address.
+///
+/// If `is_local` is true, the output address is formatted as decimal value, otherwise as hex
+/// string.
+fn print_mem_address(
+    addr: u32,
+    value: Option<Word>,
+    is_last: bool,
+    is_local: bool,
+    padding: usize,
+) {
+    if let Some(value) = value {
+        if is_last {
+            if is_local {
+                println!(
+                    "└── {addr:>5}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            } else {
+                println!(
+                    "└── {addr:#010x}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            }
+        } else if is_local {
+            println!(
+                "├── {addr:>5}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]",
+                value[0].as_int(),
+                value[1].as_int(),
+                value[2].as_int(),
+                value[3].as_int(),
+                width = padding
+            )
+        } else {
+            println!(
+                "├── {addr:#010x}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]",
+                value[0].as_int(),
+                value[1].as_int(),
+                value[2].as_int(),
+                value[3].as_int(),
+                width = padding
+            )
+        }
+    } else if is_last {
+        if is_local {
+            println!("└── {addr:>5}: EMPTY\n");
+        } else {
+            println!("└── {addr:#010x}: EMPTY\n");
+        }
+    } else if is_local {
+        println!("├── {addr:>5}: EMPTY");
     } else {
-        println!("├── {i:>2}: {}", stack[i]);
-        println!("└── ({} more items)", stack.len() - num_items);
+        println!("├── {addr:#010x}: EMPTY");
     }
 }
 
-#[cfg(not(feature = "std"))]
-fn print_vm_stack(_clk: u32, _stack: Vec<Felt>, _n: usize) {
-    // in no_std environments, this is a NOOP
+/// Returns the maximum length among the word elements.
+fn word_elem_max_len(word: Option<Word>) -> usize {
+    if let Some(word) = word {
+        word.iter().fold(0, |max, value| value.as_int().to_string().len().max(max))
```

Comment:
Rather than converting to string and counting characters, wouldn't [ilog10()](https://doc.rust-lang.org/std/primitive.u64.html#method.ilog10) give us the same result?